### DISCUSSION TITLE
[ Attestation ] How do i perform fee-less attestation/phone number linking for new users within the celo platform?

### DISCUSSION THREAD

**Discussion by GrandLegend** [celo-org/social-connect#228] [Help (Q&A)]
Komenci seems to be a good tool however I've been talked out of it due to its underdocumentation.
I'm keen to know how the celo identity protocol works and it's relationship with pin numbers used to lock accounts.
A react native example would be very handy.

 Thanks

**Comment by arthurgousset** [comment]
Adding some more context from our DM (for future reference): 

> I would like to achieve the following react native.
> 
> 1) to connect a phone number to a my address on the celo network (attestation). 
> 2) to have a pin locking the funds in my celo address/account.
> 
> I've asked around (joshc and at the valora server) also went through the valora code base to find out how it is done.
> I ran into a lib named komenci, it seemed to the the right tool only to be proven quite under documented. 
> I've been directed to you for more infor and guidance on the celo v2 identity protocol.

**Comment by arthurgousset** [comment]
Hey @GrandLegend thanks for posting the question here! 

### TLDR

- I recommend you check out @critesjosh's tutorial implementation here: https://github.com/critesjosh/register-number/blob/main/index.js
- @alecps or @eelanagaraj will be best placed to answer questions you might have about Komenci after reading through some of the context below
- I'm not sure who might be able to help with your question to "have a pin locking the funds in my celo address/account." (sorry! 🤷‍♂️). Could you be more specific? Is this simply about locking a wallet with a pin or is there some additional functionality you are looking for?

### Additional information (if you're interested)

Here is some additional context that might help put things into perspective:
- "_connecting a phone number to an address_" is currently performed by a combination of multiple _components_:
  - what we call the [Attestation Service](https://docs.celo.org/validator-guide/attestation-service), which is essentially a client that many validators run. This client is used to request/send SMS and sign various messages ("attestations").
  - a smart contract: [Attestations.sol](https://github.com/celo-org/celo-monorepo/blob/6b6ce69fde8f4868b54abd8dd267e5313c3ddedd/packages/protocol/contracts/identity/Attestations.sol), which 
    1. deals with (randomly) selecting validators that will send SMSs (this is repeated 3 times for security purposes, you can [read more about why 3 SMSs here](https://docs.celo.org/celo-codebase/protocol/identity/randomness#requesting-new-attestations) if you're interested), and 
    2. deals with storing the "connection" as you call it (or _mapping_) from phone number to Celo address. For completeness, it doesn't store a plain text phone number like `+49187...`, but a (unique) obfuscated version of the phone number (or _identifier_) e.g. `0x4589545...` that is produced using a pepper generated by ODIS, more below (or [read more here](https://docs.celo.org/celo-codebase/protocol/odis/use-cases/phone-number-privacy))
  - what we call ODIS (or **O**blivious **D**ecentralized **I**dentifier **S**ervice), which is essentially a _tool_ that takes a ([blinded](https://docs.celo.org/celo-codebase/protocol/odis#blinding), beyond the scope of this description) phone number and returns a pepper which can be used to obfuscate that plain text phone number before it's stored in [Attestations.sol](https://github.com/celo-org/celo-monorepo/blob/6b6ce69fde8f4868b54abd8dd267e5313c3ddedd/packages/protocol/contracts/identity/Attestations.sol), using
    1. a set of ODIS signer clients (or _operators_) run by various validators and other reputable parties, which sign messages, and
    2. the ODIS combiner, which in a nutshell, collects and combines the signed messages above to generate a unique pepper
  - what we call Komenci, which in a nutshell, is a tool that helps users use the above functionality when they first create a Celo account and have no transaction history (which is used to rate limit requests to ODIS) 
- most of the above functionality can be accessed today through a couple SDKs, namely @celo/contractkit, @celo/identity, and @celo/utils
  - the easiest way to wrap your head around what a _simple_ implementation might look like is @critesjosh's tutorial here: https://github.com/critesjosh/register-number/blob/main/index.js

To provide builders with the ability to "_connect phone numbers with addresses_" as you say with a different set of trade-offs (in particular easier integration), the cLabs' identity team is working on releasing tooling and smart contract(s) loosely referred to as "federated attestations" (or the _v2 identity protocol_ as @critesjosh said). Unfortunately this is not available as we speak, but we are aiming to release a non-audited version around July/August (please take these dates with a grain of salt :) )

I hope this helps provide some context for what is a somewhat complex architecture!

**Comment by GrandLegend** [comment]
@0xarthurxyz  We could not go on with this since we were using Magic as our Identity provider and therefore difficulty in getting the private key. We are working on migrating to the celo-valora way of doing it.

**Reply by arthurgousset** [reply]
Ah interesting! Could you share 2-3 lines on what made you choose [Magic](https://magic.link/) for the "identity provider"? Would love to understand where you are coming from a little more!

**Reply by steffenkrogmann** [reply]
Hi,

Chosing magic was on me.

The decision was not driven by technical factors (potentially a mistake).

We realized that most of the u we want to serve are new to crypto and they won't save their recovery phrase.

Even with all explaining in the world, they are often the kind of people who wouldn't know their email password.

Magic offered a solution to create a wallet and sign in to your wallet with your phone number (OTP).

Phone number + OTP, while not perfect, seems to be the most reliable way for many users. At least until they learn more. It is what they know.

@0xarthurxyz I heard that you guys are working on something that might address that? Scheduled for August? Could you share more?

**Reply by arthurgousset** [reply]
Hey @steffenkrogmann 👋 Thanks for the context, very useful! 
Just a brief note to say that I haven't forgotten your question above and will respond next week (if that's ok with you).
I'm OOO this week :)

**Reply by steffenkrogmann** [reply]
No worries, nothing urgent :)

A related question:
I outlined why Magic had real value for us. And I still see that as one major challenge for adoption for Celo wallets (ours, Valora, etc.).

Building that from scratch would be beyond our current internal capabilities.
But could there be a way forward to scope that a bit more together with the Celo/cLabs/Valora team? To see if that is worthwhile exploring?
I think it can be beneficial for the whole ecosystem. 

How could such a process look like?

Thank you! :)
